hnn_core/viz.py

```diff
@@ -453,27 +491,109 @@ def plot_spectrogram(dpl, *, fmin, fmax, winlen=None, tmin=None, tmax=None,
         The matplotlib figure handle.
     """
     import matplotlib.pyplot as plt
-    from scipy.signal import spectrogram
+    from scipy.signal import periodogram
+
+    # NB units_str is applied to the label, includes parentheses: ' (nAm)'
+    scaling, units = _check_scaling_units(scaling, units)
 
     sfreq = dpl.sfreq
     data, times = _get_plot_data(dpl, layer, tmin, tmax)
 
-    if winlen is None:
-        winlen = times[-1] - times[0]
-    nfft = 1e-3 * winlen * sfreq
-    nperseg = 2 ** int(np.ceil(np.log2(nfft)))
-    nperseg = min(nperseg, len(data))
+    freqs, Pxx = periodogram(data, sfreq, window='hamming', nfft=len(data))
+    if ax is None:
+        fig, ax = plt.subplots(1, 1)
+
+    # ax.plot(freqs, np.sqrt(Pxx))
+    ax.plot(freqs, Pxx)
+    if fmax is not None:
+        ax.set_xlim((fmin, fmax))
+    ax.ticklabel_format(axis='both', scilimits=(-2, 3))
+    ax.set_xlabel('Frequency (Hz)')
+    ax.set_ylabel(f'Power ({units}' + r'$^2 \ Hz^{-1}$)')
+
+    plt_show(show)
+    return ax.get_figure()
+
+
+# from mne-python, v/0.23dev0
+def _check_nfft(n, n_fft, n_per_seg, n_overlap):
+    """Ensure n_fft, n_per_seg and n_overlap make sense."""
+    if n_per_seg is None and n_fft > n:
+        raise ValueError(('If n_per_seg is None n_fft is not allowed to be > '
+                          'n_times. If you want zero-padding, you have to set '
+                          'n_per_seg to relevant length. Got n_fft of %d while'
+                          ' signal length is %d.') % (n_fft, n))
+    n_per_seg = n_fft if n_per_seg is None or n_per_seg > n_fft else n_per_seg
+    n_per_seg = n if n_per_seg > n else n_per_seg
+    if n_overlap >= n_per_seg:
+        raise ValueError(('n_overlap cannot be greater than n_per_seg (or '
+                          'n_fft). Got n_overlap of %d while n_per_seg is '
+                          '%d.') % (n_overlap, n_per_seg))
+    return n_fft, n_per_seg, n_overlap
+
+
+# inspired by mne-python (time_frequency.psd_array_welch), v/0.23dev0
+def plot_psd_welch(dpl, *, fmin=0, fmax=None, n_fft=2**14, n_overlap=0,
+                   n_per_seg=2**12, tmin=None, tmax=None, layer='agg',
+                   ax=None, show=True):
+    """Plot Power Spectral Density of dipole time course using Welch's method
+
+    Applies `~scipy.signal.welch` with ``window='hamming'``.
+
+    Parameters
+    ----------
+    dpl : instance of Dipole
+        The Dipole object.
+    fmin : float
+        Minimum frequency to plot (in Hz). Default: 0 Hz
+    fmax : float
+        Maximum frequency to plot (in Hz). Default: None (plot up to Nyquist)
```

Comment:
how do you determine the Nyquist freq? The sfreq is not in the dipole object as far as I remember

Reply:
never mind this, I just noticed it's an attribute